src/Symfony/Component/Encryption/Ciphertext.php

```diff
@@ -0,0 +1,172 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Encryption;
+
+use Symfony\Component\Encryption\Exception\DecryptionException;
+use Symfony\Component\Encryption\Exception\MalformedCipherException;
+
+/**
+ * This class is responsible for the payload API.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * @internal
+ */
+class Ciphertext implements \Stringable
+{
+    /**
+     * Algorithm used to encrypt the message.
+     */
+    private string $algorithm;
+    private string $version;
+    private string $payload;
+
+    /**
+     * Nonce used with the algorithm.
+     */
+    private string $nonce;
+
+    /**
+     * @var array<string, string> additional headers
+     */
+    private array $headers = [];
+
+    private function __construct()
+    {
+    }
+
+    /**
+     * @param array<string, string> $headers with ascii keys and values
+     */
+    public static function create(string $algorithm, string $ciphertext, string $nonce, array $headers = []): self
+    {
+        $model = new self();
+        $model->algorithm = $algorithm;
+        $model->payload = $ciphertext;
+        $model->nonce = $nonce;
+        $model->headers = $headers;
+
+        return $model;
+    }
+
+    /**
+     * Take a string representation of the ciphertext and parse it into an object.
+     *
+     * @throws MalformedCipherException
+     */
+    public static function parse(string $input): self
+    {
+        $parts = explode('.', $input);
+        if (!\is_array($parts) || 4 !== \count($parts)) {
+            throw new MalformedCipherException();
+        }
+
+        [$headersString, $payload, $nonce, $hashSignature] = $parts;
+
+        $headersString = self::base64UrlDecode($headersString);
+        $payload = self::base64UrlDecode($payload);
+        $nonce = self::base64UrlDecode($nonce);
+        $hashSignature = self::base64UrlDecode($hashSignature);
+
+        // Check if integrity hash is valid
```

Comment:
This doesn't provide integrity. HMAC-SHA256 with a secret key would provide integrity. This just tells you if an error occurred.